### PR TITLE
FLW-871 iOS: Update View validators screen

### PR DESCRIPTION
**Pull request by stas-litvinov**
- Change validators grouping
- Add error icon

### CODE REVIEW

**Review by ERussel — CHANGES_REQUESTED** [review]
Current solution doesn't take into account the state when user changes validators in the current era

File: fearless/Modules/Staking/YourValidators/View/YourValidatorTableCell.swift [hunk 1]
```diff
@@ -29,6 +29,12 @@ class YourValidatorTableCell: UITableViewCell {
         return imageView
     }()
 
+    let errorImageView: UIImageView = {
```
Comment (ERussel) [inline fearless/Modules/Staking/YourValidators/View/YourValidatorTableCell.swift:32]: I suggest to use UIStackView so we don't have holes when some status is hidden

File: fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModel.swift [hunk 2]
```diff
@@ -28,31 +28,15 @@ enum YourValidatorsViewState {
...
 }
 
 enum YourValidatorsSectionStatus {
-    case active
+    case stakeAllocated
+    case stakeNotAllocated
     case inactive
-    case waiting
-    case slashed
-    case pending
```
Comment (ERussel) [inline fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModel.swift:40]: We need to resurrect pending status

File: fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift [hunk 3]
```diff
@@ -28,52 +28,62 @@ final class YourValidatorsViewModelFactory {
...
+            shouldHaveWarning: model.stakeInfo?.oversubscribed ?? false,
+            shouldHaveError: shouldHaveError
         )
     }
 
     private func createTitle(
         for sectionStatus: YourValidatorsSectionStatus,
         count: Int
-    ) -> LocalizableResource<String> {
+    ) -> LocalizableResource<String?> {
```
Comment (ERussel) [inline fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift:48]: Looks like it is more natural to make LocalizableResource optional. There is no sense to apply localization to missing string

File: fearless/ba-RU.lproj/Localizable.strings [hunk 4]
```diff
@@ -237,7 +237,6 @@
 "staking.payouts.learn.more" = "Узнать больше про выплаты";
 "staking.reward.payout.account" = "Награда аккаунты";
 "staking.selected.accounts.mask" = "Аккаунттар(%d)";
-"staking.recommended.validators.title" = "Кәңәшләнгән валидаторҙар";
```
Comment (ERussel) [inline fearless/ba-RU.lproj/Localizable.strings:240]: Whe we need to remove localization for different screen?

Comment (stas-litvinov) [reply fearless/ba-RU.lproj/Localizable.strings:240]: This is weird, I just updated all the files from Lokalize

Comment (stas-litvinov) [reply fearless/ba-RU.lproj/Localizable.strings:240]: Will upload it to Lokalize

File: fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModel.swift (hunk 2 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModel.swift:40]: Done

File: fearless/Modules/Staking/YourValidators/View/YourValidatorTableCell.swift (hunk 1 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/YourValidators/View/YourValidatorTableCell.swift:32]: Done

File: fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift (hunk 3 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift]: Done...

File: fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift [hunk 5]
```diff
@@ -28,52 +28,62 @@ final class YourValidatorsViewModelFactory {
...
-            case .waiting:
-                return R.string.localizable.stakingYourWaitingFormat(
-                    countString,
+            }
+        }
+    }
+
+    private func createDescription(
+        for sectionStatus: YourValidatorsSectionStatus
+    ) -> LocalizableResource<String?> {
```
[the remark this answers is not in the document]
Comment (stas-litvinov) [reply fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift]: ...and done

File: fearless/ba-RU.lproj/Localizable.strings (hunk 4 above)
Comment (stas-litvinov) [reply fearless/ba-RU.lproj/Localizable.strings:240]: No, I won't upload it. Seems like it's a wrong key and we need to remove it one day

**Comment by stas-litvinov** [comment]
image
image

File: fearless/Modules/Staking/YourValidators/View/YourValidatorTableCell.swift [hunk 6]
```diff
@@ -58,38 +94,29 @@ class YourValidatorTableCell: UITableViewCell {
     }
 
     func setupLayout() {
-        contentView.addSubview(iconView)
+        contentView.addSubview(mainStackView)
+        mainStackView.snp.makeConstraints { make in
+            make.trailing.equalToSuperview()
```
Comment (ERussel) [inline fearless/Modules/Staking/YourValidators/View/YourValidatorTableCell.swift:99]: ```make.edges```?

File: fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift [hunk 7]
```diff
@@ -104,11 +138,16 @@ extension YourValidatorsViewModelFactory: YourValidatorsViewModelFactoryProtocol
...
-                    return YourValidatorsSectionStatus(modelStatus: modelStatus)
-                } else {
+                guard let modelStatus = item.myNomination else {
                     return .pending
                 }
+
+                switch modelStatus {
+                case .active:
+                    return .stakeAllocated
+                default:
```
Comment (ERussel) [inline fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift:148]:
```case .inactive => .inactive
case .waiting .slash => .stakeNotAllocated```
So we don't me any case in future

**Review by ERussel — APPROVED** [review]

**Review by ERussel — CHANGES_REQUESTED** [review]
We need to adopt our model status to match application status

File: fearless/Modules/Staking/Operations/ValidatorOperationFactory.swift [hunk 8]
```diff
@@ -104,51 +102,74 @@ final class ValidatorOperationFactory {
...
+            targetOperation: operation,
+            dependencies: [runtimeOperation] + slashingSpansWrapper.allOperations
+        )
+    }
+
+    private func createStatusesOperation(
+        for validatorIds: [AccountId],
+        electedValidatorsOperation: BaseOperation<EraStakersInfo>,
+        nominatorAddress: AccountAddress,
+        nomination _: Nomination
```
Comment (ERussel) [inline fearless/Modules/Staking/Operations/ValidatorOperationFactory.swift:133]: nomination parameter not used anymore

File: fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift [hunk 9]
```diff
@@ -100,15 +136,22 @@ final class YourValidatorsViewModelFactory {
 
 extension YourValidatorsViewModelFactory: YourValidatorsViewModelFactoryProtocol {
     func createViewModel(for model: YourValidatorsModel) throws -> [YourValidatorsSection] {
-        let allValidatos = model.currentValidators + model.pendingValidators
-        let validatorsMapping = allValidatos.reduce(
+//        let allValidatos = model.currentValidators + model.pendingValidators
```
Comment (ERussel) [inline fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift:139]: Not needed?

File: fearless/Modules/Staking/Operations/ValidatorOperationFactory.swift (hunk 8 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/Operations/ValidatorOperationFactory.swift]: fixed

File: fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift (hunk 9 above)
Comment (stas-litvinov) [reply fearless/Modules/Staking/YourValidators/ViewModel/YourValidatorsViewModelFactory.swift]: fixed

**Review by ERussel — APPROVED** [review]

**Pull request merged**